### PR TITLE
Create many-to-many association utility 

### PR DESCRIPTION
**Pull request by Rotorsoft**
## Link to Issue
Closes: #7560

## Description of Changes
- Creates utility
- Refactors the two existing xrefs (collaborations and community contracts)
- Adds missing unique constraint migration

### CODE REVIEW

**Review by burtonator — APPROVED** [review]

File: packages/commonwealth/server/migrations/20240425141332-add-unique-index-to-community-contracts.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+'use strict';
+
+module.exports = {
+  up: async (queryInterface) => {
+    await queryInterface.sequelize.transaction(async (t) => {
+      await queryInterface.sequelize.query(
+        `
+ ALTER TABLE ONLY public."CommunityContracts"
+ ADD CONSTRAINT "CommunityContracts_community_id_contract_id_key" UNIQUE(community_id, contract_id)
```
Comment (timolegros) [inline packages/commonwealth/server/migrations/20240425141332-add-unique-index-to-community-contracts.js:9]: This creates an index on `(community_id, contract_id)` so we can remove the index on `community_id` since it becomes redundant.

File: libs/model/src/models/community_contract.ts [hunk 2]
```diff
@@ -26,15 +26,13 @@ export type CommunityContractModelStatic =
...
-): CommunityContractModelStatic => {
-  const CommunityContract = <CommunityContractModelStatic>sequelize.define(
+): CommunityContractModelStatic =>
+  <CommunityContractModelStatic>sequelize.define<CommunityContractInstance>(
     'CommunityContract',
     {
       id: { type: dataTypes.INTEGER, primaryKey: true, autoIncrement: true },
       community_id: { type: dataTypes.STRING, allowNull: false },
       contract_id: { type: dataTypes.INTEGER, allowNull: false },
-      created_at: { type: dataTypes.DATE, allowNull: false },
```
Comment (timolegros) [inline libs/model/src/models/community_contract.ts:36]: why is this removed?

**Review by timolegros — CHANGES_REQUESTED** [review]

File: libs/model/src/models/community_contract.ts (hunk 2 above)
Comment (Rotorsoft) [reply libs/model/src/models/community_contract.ts:36]: sequelize creates those fields automatically when when `timestamps` = true

Comment (timolegros) [reply libs/model/src/models/community_contract.ts:36]: Yes but so far the pattern has been we still define them explicitly. Do we not want to do that anymore?

File: libs/model/src/models/community_contract.ts [hunk 3]
```diff
@@ -26,36 +26,19 @@ export type CommunityContractModelStatic =
...
   dataTypes: typeof DataTypes,
-): CommunityContractModelStatic => {
-  const CommunityContract = <CommunityContractModelStatic>sequelize.define(
+): CommunityContractModelStatic =>
+  <CommunityContractModelStatic>sequelize.define<CommunityContractInstance>(
     'CommunityContract',
     {
       id: { type: dataTypes.INTEGER, primaryKey: true, autoIncrement: true },
       community_id: { type: dataTypes.STRING, allowNull: false },
       contract_id: { type: dataTypes.INTEGER, allowNull: false },
```
Comment (timolegros) [inline libs/model/src/models/community_contract.ts:35]: The unique constraint added in the migration should be defined in the model here as well.

File: libs/model/src/models/community_contract.ts (hunk 2 above)
Comment (Rotorsoft) [reply libs/model/src/models/community_contract.ts:36]: that's fine, but I added the generic type to define, and it complains b/c those attributes are not in the type

define<CommunityContractInstance>

**Review by timolegros — APPROVED** [review]
